### PR TITLE
Slicing parameters - large VRAM savings

### PR DESCRIPTION
**Pull request by dxqbYD**
As you know, Prodigy uses 4 internal states for each training parameter, resulting in a large VRAM footprint. AdamW in comparison uses 2 states per parameter.

These additional 2 states, p0 and s, are used to calculate d and estimate the learning rate. 

Experiments show that it is not necessary (for large models) to store p0 and s for each and every training parameter. An equally good estimation of the learning rate can be achieved by only using a fraction of the training parameters.

This PR introduces a new hyperparameter "slice_p". A value of 10 only uses every 10th training parameter for estimating the learning rate.
This has been tested using OneTrainer on the following models, with a significant impact on VRAM requirements:

| Model | Prodigy | Prodigy sliced | AdamW |
|--------|--------|--------|--------|
| Flux Dev LoRA Rank 128 | 12,1 GB | **10,7 GB** | 10.5 GB |
| SDXL full unet finetune | 31,9 GB | **22,6 GB** | 21,7 GB |

LR estimations are very similar between 'Prodigy' and 'Prodigy sliced'; so are training results.
slice_p was 10. VRAM values given are total VRAM values used by OneTrainer, not only Prodigy overhead.

Note: the default for slice_p in this PR is 1, resulting in no change!
Set a higher value if you want to test this PR.

### CODE REVIEW

**Comment by adefazio** [comment]
This seems like a good idea! Konstantin and I will take a look.

**Comment by Arcitec** [comment]
This looks extremely interesting. Combined with #23 it seems like Prodigy could finally become as good as AdamW's VRAM usage. :exploding_head:

**Comment by dxqbYD** [comment]
I have ran some more tests overnight that look very promising.

Unfortunately, running Prodigy on OneTrainer isn't deterministic, so I cannot show that Prodigy finds identical results with slicing. But the following does show that the results are equally good.

All tests with Flux Dev LoRA, with some hyperparameter variation that usually affect the LR you want:

Batch size 1, rank 128, alpha 1
grafik

**Black** and **blue** are without slicing, **pink** and **orange** is sliced.
Note that the random difference between the two non-sliced runs in black and blue (which would be identical in a deterministic testcase) seems bigger than the difference between sliced and non-sliced.

Batch size 2, rank 128, alpha 1
grafik


Batch size 1, rank 128, alpha 128:
grafik

Black seems like a relatively large outlier late during training, but again note that blue, with identical parameters and data, and both unsliced, does not have that; probably random. A higher alpha might cause Prodigy to estimate lower at the start, but in both sliced and unsliced.

Batch size 1, rank 16, alpha 1:
grafik

Last training run in orange was cut short, because I needed my GPU for something else :)

**Comment by dxqbYD** [comment]
Optimizer settings for all tests:
grafik

slice_p was 10 for the orange and pink, 1 for black and blue

**Comment by adefazio** [comment]
As long as it gives LR values in the same ball-park it will be useful. I wouldn't expect identical values.

**Comment by dxqbYD** [comment]
The logarithmic scale of tensorboard seems to be broken. Even the largest outliers above are actually quite close together, not at 5e-4:
grafik

**Comment by Arcitec** [comment]
Okay that actually looks amazing. They're very close despite -45% reduction in VRAM usage! And being a controllable setting (defaulted to no slicing), I don't see any drawbacks to this awesome improvement! :)

**Comment by adefazio** [comment]
It might make sense to use a odd or prime number as the slice value so that the p value doesn't accidentally result in skipping certain tensor channels if they happen to be a multiple of p. Maybe we could recommend p=11 in the docs? 
I'm kicking off some of the experiments from the original prodigy paper using p=11 to see if we get good results.

**Comment by adefazio** [comment]
Let me know if p=11 also works in your experiments, I will merge for now with that recommendation.

**Comment by dxqbYD** [comment]
> Let me know if p=11 also works in your experiments, I will merge for now with that recommendation.

did the original experiments of the paper have good results with slice_p==11?

It's a good theory that 11 might be better than 10 in case the parameters sliced at 10 are more uniform than at 11. But I don't think it makes sense to re-run my experiments, because they already had equally good results at 10 compared to 1.

*If* there is a case that doesn't work at 10, 11 might be an improvement.
Doesn't hurt in any case to recommend 11.

**Comment by adefazio** [comment]
I have ran the DLRM experiment and it appears to be working as before. I'll run further experiments as well but since this is not a breaking change I'm going to merge in for now so the other PRs can merge on top of this.

**Comment by konstmish** [comment]
Thanks for the PR! I suspected that one could store p0 and s with low precision but I didn't expect we could just skip the parameters and it would work equally well on most problems. Great work!

**Comment by konstmish** [comment]
@adefazio I updated the readme file to add instructions on using the new hyperparameter, please take a look.

**Comment by Arcitec** [comment]
Thanks everyone for working on this and confirming that it works. What a great improvement! :) Can finally use Prodigy on consumer GPUs.

**Comment by Arcitec** [comment]
@dxqbYD Confirmed working perfectly for me.

I just remembered one thing though: Slicing creates a view to the original data, keeping the original data in memory. Have all the detach/clone calls been verified to ensure that they copy the data into new, contiguous memory so the original tensors can be released from memory? In other words, be sure it's not just views, so that the original data can be garbage collected.

**Comment by konstmish** [comment]
> I didn't know at the time whether it was "legal" to change the gradient buffer in the optimizer, but apparently it is, because the original authors have done the same in their schedule-free repo?
Technically it's legal, but it may also lead to some unexpected behaviour on the user side in case they track gradient norm or do something with the gradients outisde of the optimizer. That being said, gradient clipping already modifies the gradient, so it's not too bad to change it inside the optimizer. For ScheduleFree methods I think it was important because of the benchmark that we were testing on (was it because of the Criteo task @adefazio?), but in most other cases the extra memory required to create this buffer is probably not significant.

**Comment by adefazio** [comment]
Changing the grad buffer is ok but as Konstantin said it will sometimes break user code if they don't expect it. I haven't heard any complaints in the schedule-free case.

In terms of the detaches and .data calls, they should all be removed at some point and the code changed to use @torch.no_grad(), the old style is depreciated. I've already done this for the schedule-free codebase. It's not pressing though, no copying is done when you call detach or .data.

**Comment by Arcitec** [comment]
@dxqbYD The concern is this:

image

That's how slices work:

https://pytorch.org/docs/stable/tensor_view.html

They don't allocate new memory. They just create a small metadata structure which says "read from original Tensor's memory location, skip every X values".

So original data (what we sliced from) stays allocated in memory!

---

Apparently it's possible to check if a view still points at the original tensor's full data by looking at the memory pointers:

```py
sliced_tensor.storage().data_ptr() == original_tensor.storage().data_ptr()
```

---

The original data cannot be garbage collected while views/slices still point at it.

To release the original data we must make sure the slice becomes a copy, not a view-reference anymore.

There are several ways to ensure it becomes a copy:

- Call [t.contiguous()](https://pytorch.org/docs/stable/generated/torch.Tensor.contiguous.html#torch.Tensor.contiguous) immediately after slicing. What this does for us: If we call slice `[::1]` we are already slicing into the exact original data, so `contiguous` just immediately returns the original tensor (not a copy; but in this case we don't want a copy since we want the original data, so that's great). Else, if we used a larger stride `[::2]` or higher, it will copy the data into a new tensor.
- Call [t.clone()](https://pytorch.org/docs/stable/generated/torch.Tensor.clone.html#torch.Tensor.clone) which copies the sliced view data into a real, standalone tensor. Downside is that if we did a `[::1]` slice we end up copying the entire original data a 2nd time. So `contiguous()` looks more appropriate.

**Comment by dxqbYD** [comment]
I don't see the issue. Views are okay, we don't want copies except at state initialization.
Look at the code which tensors are actually sliced during runtime - it's pre-existing tensors that aren't created for the purpose of slicing. I don't expect there to be any allocations because of that.

The (p0-p) tensor was always a temporary tensor that is discarded immediatly after, even before slicing. I expect that it will be collected as before.

**Comment by Arcitec** [comment]
@dxqbYD Yeah views are fine. We should just ensure that no memory stays allocated anywhere (VRAM or RAM) because of them. :)

**Comment by konstmish** [comment]
Yep, thanks for the reminder!

**Comment by dxqbYD** [comment]
it's merged by OneTrainer now, in case you prefer some images once in a while over of your academic benchmark scores :)

Happy New Year!
